# Assignment 1

Huzaifa Rehman,
24I-2554, 
DS(A)

In [16]:
import pandas as pd
import numpy as np
import sqlite3

df = pd.read_csv('adult.csv')
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


## Task 1 

In [17]:
df.info()
print('Before replacing ?:')
print(df.isnull().sum())

df.replace('?', np.nan, inplace=True)

print('After replacing ?:')
print(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
Before replacing ?:
age               0
workclass         0
fnlwgt            0
education    

## Task 2 

In [18]:
missing = pd.DataFrame({'Missing Count': df.isnull().sum(), 'Missing Percentage': (df.isnull().sum() / len(df)) * 100})
missing.sort_values('Missing Count', ascending=False)

,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486
fnlwgt,0,0.000000
education,0,0.000000
education.num,0,0.000000
age,0,0.000000
marital.status,0,0.000000
relationship,0,0.000000
sex,0,0.000000


## Task 3 

Missing values are categorical, so they are filled with Unknown. No numeric column has missing values.

In [19]:
df['workclass'].fillna('Unknown', inplace=True)
df['occupation'].fillna('Unknown', inplace=True)
df['native.country'].fillna('Unknown', inplace=True)
print(df.isnull().sum())

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


C:\Users\SART\AppData\Local\Temp\ipykernel_7100\2829498595.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['workclass'].fillna('Unknown', inplace=True)
C:\Users\SART\AppData\Local\Temp\ipykernel_7100\2829498595.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example,

## Task 4
Only exact duplicates are removed. Duplicates without the income column are kept because different people can have the same attributes but different income labels.

In [20]:
print('Exact duplicates:', df.duplicated().sum())
print('Duplicates without income:', df.drop(columns=['income']).duplicated().sum())

df = df.drop_duplicates()
print('Shape after removing exact duplicates:', df.shape)

Exact duplicates: 24
Duplicates without income: 25
Shape after removing exact duplicates: (32537, 15)


## Task 5 

In [21]:
print('Education:')
print(df['education'].unique())
print('\nMarital Status:')
print(df['marital.status'].unique())
print('\nNative Country:')
print(df['native.country'].unique())

for column in df.select_dtypes(include='object').columns:
    df[column] = df[column].str.strip()

Education:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Marital Status:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Native Country:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


## Task 6 

In [22]:
print('Numeric columns:')
display(df.describe())
print('Categorical columns:')
display(df.describe(include='object'))
print('Mean and median:')
print(df[['age', 'hours.per.week']].agg(['mean', 'median']))

Numeric columns:


,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32537.000000,3.253700e+04,32537.000000,32537.000000,32537.000000,32537.000000
mean,38.585549,1.897808e+05,10.081815,1078.443741,87.368227,40.440329
std,13.637984,1.055565e+05,2.571633,7387.957424,403.101833,12.346889
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.369930e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


Categorical columns:


,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32537,32537,32537,32537,32537,32537,32537,32537,32537
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22673,10494,14970,4136,13187,27795,21775,29153,24698


Mean and median:
              age  hours.per.week
mean    38.585549       40.440329
median  37.000000       40.000000


Mean age is slightly higher than median age. Mean hours per week and median hours per week are very close.

## Task 7 

In [23]:
print('Most common occupation:')
print(df['occupation'].value_counts().head(1))
print('\nSex percentage:')
print(df['sex'].value_counts(normalize=True) * 100)
print('\nIncome percentage:')
print(df['income'].value_counts(normalize=True) * 100)

Most common occupation:
occupation
Prof-specialty    4136
Name: count, dtype: int64

Sex percentage:
sex
Male      66.92381
Female    33.07619
Name: proportion, dtype: float64

Income percentage:
income
<=50K    75.907428
>50K     24.092572
Name: proportion, dtype: float64


The income classes are imbalanced because <=50K has many more records than >50K.

## Task 8 

In [24]:
education_check = df.groupby('education')['education.num'].unique()
print(education_check)

inconsistent = education_check[education_check.apply(len) > 1]
print('\nInconsistencies:')
print(inconsistent)

if len(inconsistent) == 0:
    print('No inconsistencies found.')

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object

Inconsistencies:
Series([], Name: education.num, dtype: object)
No inconsistencies found.


## Task 9 

In [25]:
raw_df = pd.read_csv('adult.csv')
connection = sqlite3.connect(':memory:')
raw_df.to_sql('adult_income', connection, index=False, if_exists='replace')

sql_df = pd.read_sql_query('SELECT * FROM adult_income', connection)
age_over_30 = pd.read_sql_query('SELECT * FROM adult_income WHERE age > 30', connection)

print('Original CSV shape:', raw_df.shape)
print('SQL SELECT * shape:', sql_df.shape)
print('Age > 30 shape:', age_over_30.shape)
display(age_over_30.head())
connection.close()

Original CSV shape: (32561, 15)
SQL SELECT * shape: (32561, 15)
Age > 30 shape: (21989, 15)


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [26]:
df.to_csv('adult_cleaned.csv', index=False)
print('adult_cleaned.csv saved successfully')

adult_cleaned.csv saved successfully


## Task 10

- Dataset: 32,561 rows and 15 columns before cleaning.
- Missing values: occupation, workclass, and native.country contained ? values.
- Duplicates: exact duplicate rows were removed.
- Inconsistent data: whitespace was removed from categorical columns.
- Key observation: <=50K is much more common than >50K, so the income classes are imbalanced.
- Cleaning: missing categorical values were filled with Unknown; no numeric values were missing.